File: rust/src/medium/affine_cipher.rs
#[derive(Debug, Eq, PartialEq)]
pub enum AffineCipherError {
    NotCoprime(i32),
}

pub fn encode(plaintext: &str, a: i32, b: i32) -> Result<String, AffineCipherError> {
    if !is_coprime(a, 26) {
        return Err(AffineCipherError::NotCoprime(a));
    }

    Ok(plaintext
        .chars()
        .filter(|ch| ch.is_alphanumeric())
        .map(|ch| encode_char(ch, a, b))
        .enumerate()
        .map(|(ind, ch)| match ind > 0 && ind % 5 == 0 {
            true => format!(" {}", ch),
            false => ch.to_string(),
        })
        .collect())
}

pub fn decode(ciphertext: &str, a: i32, b: i32) -> Result<String, AffineCipherError> {
    if !is_coprime(a, 26) {
        return Err(AffineCipherError::NotCoprime(a));
    }

    Ok(ciphertext
        .chars()
        .filter(|ch| ch.is_alphanumeric())
        .map(|ch| decode_char(ch, a, b))
        .collect())
}

fn is_coprime(a: i32, m: i32) -> bool {
    match a {
        0 => a == 1 || m == 1,
        _ => match a > m {
            true => is_coprime(a.rem_euclid(m), a),
            false => is_coprime(m % a, a),
        },
    }
}

fn encode_char(ch: char, a: i32, b: i32) -> char {
    let alphabet = ('a'..='z').enumerate();

    if char::is_digit(ch, 10) {
        return ch;
    }

    let index = alphabet
        .clone()
        .find(|(_a, b)| *b == ch.to_ascii_lowercase())
        .map(|(a, _b)| a)
        .expect("unknown char");

    let shift = (a as usize * index + b as usize) % 26;

    alphabet
        .clone()
        .find(|(ind, _b)| *ind == shift)
        .map(|(_a, b)| b)
        .expect("unknown index")
}

fn modular_multiplicative_inverse(a: i32) -> i32 {
    let m = 26;

    (1..m).find(|n| (n * a) % m == 1).unwrap()
}

fn decode_char(ch: char, a: i32, b: i32) -> char {
    let alphabet = ('a'..='z').enumerate();

    if char::is_digit(ch, 10) {
        return ch;
    }

    let index = alphabet
        .clone()
        .find(|(_a, b)| *b == ch.to_ascii_lowercase())
        .map(|(a, _b)| a)
        .expect("unknown char");

    let shift = (modular_multiplicative_inverse(a) * (index as i32 - b)).rem_euclid(26);
    dbg!(shift);

    alphabet
        .clone()
        .find(|(ind, _b)| *ind == shift as usize)
        .map(|(_a, b)| b)
        .expect("unknown index")
}

#[test]
fn encode_yes() {
    assert_eq!(encode("yes", 5, 7).unwrap(), "xbt")
}

#[test]
fn encode_all_the_letters() {
    assert_eq!(
        encode("The quick brown fox jumps over the lazy dog.", 17, 33).unwrap(),
        "swxtj npvyk lruol iejdc blaxk swxmh qzglf"
    )
}

#[test]
fn encode_numbers() {
    assert_eq!(
        encode("Testing,1 2 3, testing.", 3, 4).unwrap(),
        "jqgjc rw123 jqgjc rw"
    )
}

#[test]
fn decode_exercism() {
    assert_eq!(decode("tytgn fjr", 3, 7).unwrap(), "exercism")
}

#[test]
fn decode_with_a_not_coprime_to_m() {
    const EXPECTED_ERROR: AffineCipherError = AffineCipherError::NotCoprime(13);

    match decode("Test", 13, 11) {
        Err(EXPECTED_ERROR) => (),

        Err(e) => panic!(
            "Incorrect error: expected: {:?}, actual: {:?}.",
            EXPECTED_ERROR, e
        ),

        Ok(r) => panic!(
            "Cannot encode/decode when a is coprime to m: expected: {:?}, actual: {:?}.",
            EXPECTED_ERROR, r
        ),
    }
}

File: elixir/lib/medium/affine_cipher.ex
defmodule AffineCipher do
  @alphabet_length 26
  @encoded_word_length 5
  @digits ?0..?9

  def encode(%{a: a, b: b}, message) do
    if coprime?(a) do
      {:ok, do_encode(a, b, message)}
    else
      {:error, "a and m must be coprime."}
    end
  end

  defp do_encode(a, b, message) do
    message
    |> normalize()
    |> String.to_charlist()
    |> Stream.map(&encode_char(&1, a, b))
    |> Enum.chunk_every(@encoded_word_length)
    |> Enum.join(" ")
  end

  defp coprime?(a) do
    Integer.gcd(a, @alphabet_length) == 1
  end

  defp normalize(str) do
    str
    |> String.downcase()
    |> String.replace(~r/[^\d\w]/, "")
  end

  defp encode_char(ch, _a, _b) when ch in @digits, do: ch

  defp encode_char(ch, a, b) do
    index = ch - ?a
    shift = rem(a * index + b, @alphabet_length)

    ?a + shift
  end

  def decode(%{a: a, b: b}, encrypted) do
    if coprime?(a) do
      {:ok, do_decode(a, b, encrypted)}
    else
      {:error, "a and m must be coprime."}
    end
  end

  defp do_decode(a, b, message) do
    mmi = modular_multiplicative_inverse(a)

    message
    |> normalize()
    |> String.to_charlist()
    |> Enum.map(&decode_char(&1, mmi, b))
    |> List.to_string()
  end

  defp modular_multiplicative_inverse(a) do
    Enum.find(1..@alphabet_length, fn n -> rem(n * a, @alphabet_length) == 1 end)
  end

  defp decode_char(ch, _mmi, _b) when ch in @digits, do: ch

  defp decode_char(ch, mmi, b) do
    index = ch - ?a
    shift = Integer.mod(mmi * (index - b), 26)

    ?a + shift
  end
end
